jcl/src/java.base/share/classes/java/lang/invoke/MethodHandle.java

```diff
@@ -853,110 +853,111 @@ private static final Object resolveConstantDynamic(long j9class, String name, St
-					}
-					break;
+		/* Check if we need to treat the last parameter specially when handling primitives.
+		 * The type of the varargs array will determine how primitive ints from the constantpool
+		 * get boxed: {Boolean, Byte, Short, Character or Integer}.
+		 */
+		boolean treatLastArgAsVarargs = bsm.isVarargsCollector();
+		Class<?> varargsComponentType = bsm.type.lastParameterType().getComponentType();
+		int bsmTypeArgCount = bsm.type.parameterCount();
+
+		for (int i = 0; i < bsmArgCount; i++) {
+			int staticArgIndex = BSM_OPTIONAL_ARGUMENTS_START_INDEX + i;
+			short index = UNSAFE.getShort(bsmArgs + (i * BSM_ARGUMENT_SIZE));
+			int cpType = getCPTypeAt(internalRamClass, index);
+			Object cpEntry = null;
+			switch (cpType) {
+			case 1:
+				cpEntry = cp.getClassAt(index);
+				if (cpEntry == null) {
+					throw throwNoClassDefFoundError(classObject, index);
 				}
-				case 4:
-					cpEntry = cp.getFloatAt(index);
-					break;
-				case 5:
-					cpEntry = cp.getLongAt(index);
-					break;
-				case 6:
-					cpEntry = cp.getDoubleAt(index);
-					break;
-				case 13:
-					cpEntry = getCPMethodTypeAt(internalRamClass, index);
-					break;
-				case 14:
-					cpEntry = getCPMethodHandleAt(internalRamClass, index);
-					break;
-				case 17:
-					cpEntry = getCPConstantDynamicAt(internalRamClass, index);
-					break;
-				default:
-					// Do nothing. The null check below will throw the appropriate exception.
+				break;
+			case 2:
+				cpEntry = cp.getStringAt(index);
+				break;
+			case 3: {
+				int cpValue = cp.getIntAt(index);
+				Class<?> argClass;
+				if (treatLastArgAsVarargs && (staticArgIndex >= (bsmTypeArgCount - 1))) {
+					argClass = varargsComponentType;
+				} else {
+					argClass = bsm.type.parameterType(staticArgIndex);
 				}
-				
-				cpEntry.getClass();	// Implicit NPE
-				staticArgs[staticArgIndex] = cpEntry;
+				if (argClass == Short.TYPE) {
+					cpEntry = (short) cpValue;
+				} else if (argClass == Boolean.TYPE) {
+					cpEntry = cpValue == 0 ? Boolean.FALSE : Boolean.TRUE;
+				} else if (argClass == Byte.TYPE) {
+					cpEntry = (byte) cpValue;
+				} else if (argClass == Character.TYPE) {
+					cpEntry = (char) cpValue;
+				} else {
+					cpEntry = cpValue;
+				}
+				break;
+			}
+			case 4:
+				cpEntry = cp.getFloatAt(index);
+				break;
+			case 5:
+				cpEntry = cp.getLongAt(index);
+				break;
+			case 6:
+				cpEntry = cp.getDoubleAt(index);
+				break;
+			case 13:
+				cpEntry = getCPMethodTypeAt(internalRamClass, index);
+				break;
+			case 14:
+				cpEntry = getCPMethodHandleAt(internalRamClass, index);
+				break;
+			case 17:
+				cpEntry = getCPConstantDynamicAt(internalRamClass, index);
+				break;
+			default:
+				// Do nothing. The null check below will throw the appropriate exception.
 			}
 
+			cpEntry.getClass();	// Implicit NPE
+			staticArgs[staticArgIndex] = cpEntry;
+		}
+
+		/* JVMS JDK11 5.4.3.6 Dynamically-Computed Constant and Call Site Resolution */
```

Comment:
```
/* JVMS JDK11 5.4.3.6 Dynamically-Computed Constant and Call Site Resolution:
 * requires that only exceptions thrown by the BSM are wrapped in a BootstrapMethodError.
 * Note, subclasses of Error should not be wrapped.
* Exceptions thrown by earlier parts of resolution should be passed through unwrapped.
*/
```

Something like this clearly lays out the rules as they were understood at the time the code was written.  (Please double check I've expressed them correctly).

This makes it easier to know what the code should be doing when debugging something or trying to understand why a particular exception was wrapped / not.